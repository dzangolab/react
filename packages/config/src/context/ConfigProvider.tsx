import React, { createContext } from "react";

import { AppConfig } from "..";

interface Properties {
  children: React.ReactNode;
  appConfig: AppConfig;
}

const configContext = createContext<AppConfig | undefined>(undefined);

const ConfigProvider = ({ children, appConfig }: Properties) => {
  return (
    <configContext.Provider value={appConfig}>
      {children}
    </configContext.Provider>
  );
};

export default ConfigProvider;
export { configContext };
