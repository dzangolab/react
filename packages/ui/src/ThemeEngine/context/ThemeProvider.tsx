import { createContext, useContext, useState } from "react";

import { getThemeConfig } from "../theme-config";

import type { DzTheme } from "../types";

const ThemeContext = createContext<{
  theme: DzTheme;
  setTheme: (theme: DzTheme) => void;
}>({
  theme: "auto",
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  setTheme: () => {},
});

export const DzThemeProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const [theme, setTheme] = useState<DzTheme>(getThemeConfig().defaultTheme);

  return (
    <ThemeContext.Provider value={{ theme, setTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useDzThemeContext = () => {
  return useContext(ThemeContext);
};
