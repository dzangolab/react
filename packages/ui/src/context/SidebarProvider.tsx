import React, { createContext, useMemo, useState } from "react";

interface Properties {
  children: React.ReactNode;
}

interface SidebarContextProperties {
  open: boolean;
  toggleSidebar: () => void;
}

const sidebarContext = createContext<SidebarContextProperties | undefined>(
  undefined
);

const SidebarProvider = ({ children }: Properties) => {
  const [open, setOpen] = useState<boolean>(true);

  const toggleSidebar = () => {
    setOpen((previous) => !previous);
  };

  const providerValue = useMemo(() => ({ open, toggleSidebar }), [open]);

  return (
    <sidebarContext.Provider value={providerValue}>
      {children}
    </sidebarContext.Provider>
  );
};

export default SidebarProvider;
export { sidebarContext };
export type { SidebarContextProperties };
