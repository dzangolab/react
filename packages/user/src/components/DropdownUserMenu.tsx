import { useTranslation } from "@dzangolab/react-i18n";
import { useId, useState, useEffect, useRef } from "react";
import { toast } from "react-toastify";

import DropdownUserMenuItem from "./DropdownUserMenuItem";
import { removeUserData } from "../helpers";
import { useUser } from "../hooks";
import logout from "../supertokens/logout";
import { UserMenuItemType } from "../types";

import "../assets/css/dropdownUserMenu.css";

interface Properties {
  userMenu?: UserMenuItemType[];
}

const DropdownUserMenu: React.FC<Properties> = ({ userMenu }) => {
  const id = useId();
  const { user, setUser } = useUser();
  const [expanded, setExpanded] = useState(false);
  const { t } = useTranslation("user");
  const navBarReference = useRef<HTMLElement | null>(null);

  const signout = async () => {
    if (await logout()) {
      await removeUserData();
      setUser(undefined);
      toast.success(t("logout.message"));
    }
  };

  const collapseItems = () => {
    setExpanded(!expanded);
  };

  const signoutRoute = {
    name: "userMenu.logout",
    onClick: signout,
    route: undefined,
  };
  const profileRoute = {
    name: "userMenu.profile",
    route: "/profile",
    onClick: undefined,
  };

  const fallbackItems = [profileRoute, signoutRoute];
  const menuItems = userMenu ? [...userMenu, ...fallbackItems] : fallbackItems;

  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (
        navBarReference.current &&
        !navBarReference.current.contains(event.target as Node)
      ) {
        setExpanded(false);
      }
    }

    document.addEventListener("click", handleClickOutside);

    return () => {
      document.removeEventListener("click", handleClickOutside);
    };
  }, []);

  return (
    <nav
      ref={navBarReference}
      className={`user-menu ${expanded ? "expanded" : ""}`}
    >
      <div className="email" onClick={() => setExpanded(!expanded)}>
        {user?.email}
        <span className={"toggle"}>&#9662;</span>
      </div>

      {expanded && (
        <ul className="dropdown">
          {menuItems.map(({ name, onClick, route }) => (
            <DropdownUserMenuItem
              onClick={() => {
                onClick && onClick();
                collapseItems();
              }}
              route={route}
              key={`${id}__${name}`}
            >
              {t(name)}
            </DropdownUserMenuItem>
          ))}
        </ul>
      )}
    </nav>
  );
};

export default DropdownUserMenu;
