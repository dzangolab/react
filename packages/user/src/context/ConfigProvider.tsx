import React, { createContext } from "react";

import { UserConfig } from "@/types/config";

interface Properties {
  children: React.ReactNode;
  config: UserConfig;
}

const configContext = createContext<UserConfig | undefined>(undefined);

const ConfigProvider = ({ children, config }: Properties) => {
  return (
    <configContext.Provider value={config}>{children}</configContext.Provider>
  );
};

export default ConfigProvider;

export { configContext };
