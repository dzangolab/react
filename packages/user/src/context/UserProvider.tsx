import React, { createContext, useEffect, useState } from "react";

import { getUserData, setUserData } from "../helpers";
import { useConfig } from "../hooks";
import {
  getUserRoles,
  isUserVerified,
  verifySessionRoles,
} from "../supertokens/helpers";
import { UserContextType, UserType } from "../types";
import { isEmailVerificationEnabled } from "../utils/emailVerificationconfig";

interface Properties {
  children: React.ReactNode;
}

const userContext = createContext<UserContextType | undefined>(undefined);

const UserProvider = ({ children }: Properties) => {
  const [user, setUser] = useState<UserType | undefined>(undefined);
  const [loading, setLoading] = useState<boolean>(true);
  const appConfig = useConfig();

  useEffect(() => {
    const getUser = async () => {
      try {
        if (
          appConfig &&
          (await verifySessionRoles(appConfig.user.supportedRoles))
        ) {
          const userInfo = await getUserData();

          if (userInfo) {
            setUser(userInfo);
          }
        }
      } catch (error) {
        //
      } finally {
        setLoading(false);
      }
    };

    getUser();
  }, []);

  const setVerifiedUserWithRoles = async (user: UserType | undefined) => {
    let roles, isEmailVerified;

    if (user) {
      roles = await getUserRoles();
      if (isEmailVerificationEnabled(appConfig)) {
        isEmailVerified = await isUserVerified();
        await setUserData({
          ...user,
          roles: roles,
          isEmailVerified: isEmailVerified,
        });
        setUser({ ...user, roles: roles, isEmailVerified: isEmailVerified });
      } else {
        await setUserData({
          ...user,
          roles: roles,
        });
        setUser({ ...user, roles: roles });
      }
    } else {
      setUser(undefined);
    }
  };

  return (
    <userContext.Provider
      value={{ user, loading, setUser: setVerifiedUserWithRoles }}
    >
      {loading ? null : children}
    </userContext.Provider>
  );
};

export default UserProvider;
export { userContext };
