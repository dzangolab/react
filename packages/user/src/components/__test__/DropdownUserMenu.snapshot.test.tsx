import React from "react";
import { create } from "react-test-renderer";
import { expect, test, vi } from "vitest";

import { userContext } from "../../context/UserProvider";
import DropdownUserMenu from "../DropdownUserMenu";

import type {
  ReactTestRenderer,
  ReactTestRendererJSON,
} from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";

function toJson(component: ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const values = {
    setUser: vi.fn(),
    user: {
      id: "5",
      email: "test@gmail.com",
      timeJoined: 320,
      profile: null,
      roles: ["USER"],
    },
    loading: false,
  };

  const component = create(
    <MemoryRouter>
      <userContext.Provider value={values}>
        <DropdownUserMenu />
      </userContext.Provider>
    </MemoryRouter>
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
