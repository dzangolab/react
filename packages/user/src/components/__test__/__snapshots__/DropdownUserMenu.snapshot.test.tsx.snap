// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<div>
  <nav
    className="user-menu "
  >
    <div
      className="email"
      onClick={[Function]}
    >
      test@gmail.com
      <span
        className="toggle"
      >
        â–¾
      </span>
    </div>
  </nav>
</div>
`;
